### PR TITLE
Add Phi to the DNN + Displaced Track Weighting

### PR DESCRIPTION
**Pull request by GNiendorf**
Slides: [PR_114.pdf](https://github.com/user-attachments/files/17652119/PR_114.pdf)

### **Overview**

Using different weights for specific classes that are considered important is a common technique in ML. Displaced tracks make up a small fraction of the training set, but weighting them higher in the loss function encourages the DNN to learn features from the Phi input variables that don't negatively impact the performance on displaced tracks. The tradeoff is that the DNN learns features beneficial to displaced tracks, representing only a small fraction of all tracks, in exchange for a slight decrease in performance on the overall training dataset. However, the added Phi variables to the DNN still lead to a large improvement in overall performance, which you can see in the training ipynb (going from AUC=0.924->0.939).

I'll post my slides after Tuesday, but here is an example from 2x weight on the cube sample.
Screenshot 2024-11-03 at 10 54 29 PM

DNN Predictions for displaced tracks with no additional weighting.
DNN Predictions for displaced tracks with no weighting

DNN Predictions for displaced tracks at 4x weighting (this current PR).
DNN Score for Displaced T5s

Current FR.
Screenshot 2024-11-03 at 10 38 49 AM

This PR's Current FR (at 4x disp weighting).
Screenshot 2024-11-03 at 10 37 56 AM

### CODE REVIEW

**Comment by slava77** [comment]
> This PR's Current FR (at 4x disp weighting).

Barrel is getting better and better, very nice.

What's happening in the endcaps (eta>1): pT5 fakes barely move. One possibility is that we are just attaching wrong pLS to T5s, or pLSs are fake; another is that T5s are fake.
For the latter it may be useful to revive the lower level object performance plots (before they get combined in TCs).

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h [hunk 1]
```diff
@@ -58,6 +58,12 @@ namespace lst::t5dnn {
     float eta4 = alpaka::math::abs(acc, mdsInGPU.anchorEta[mdIndex4]);  // outer T3 anchor hit 4 eta (t3_2_eta)
     float eta5 = alpaka::math::abs(acc, mdsInGPU.anchorEta[mdIndex5]);  // outer T3 anchor hit 5 eta (t3_4_eta)
 
+    float phi1 = alpaka::math::abs(acc, mdsInGPU.anchorPhi[mdIndex1]);  // inner T3 anchor hit 1 phi (t3_0_phi)
+    float phi2 = alpaka::math::abs(acc, mdsInGPU.anchorPhi[mdIndex2]);  // inner T3 anchor hit 2 phi (t3_2_phi)
+    float phi3 = alpaka::math::abs(acc, mdsInGPU.anchorPhi[mdIndex3]);  // inner T3 anchor hit 3 phi (t3_4_phi)
+    float phi4 = alpaka::math::abs(acc, mdsInGPU.anchorPhi[mdIndex4]);  // outer T3 anchor hit 4 phi (t3_2_phi)
+    float phi5 = alpaka::math::abs(acc, mdsInGPU.anchorPhi[mdIndex5]);  // outer T3 anchor hit 5 phi (t3_4_phi)
```
Comment (slava77) [inline RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: it seems that it would be useful to get signed deltaPhi before computing the abs. To at least avoid a wrap-around mess

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: Oh yeah, good point. Thank you

Comment (slava77) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: Looking at the CI plots, I didn't find any striking pathologies in plots vs phi for efficiency.
although there are two bins near 0 and π in T5 fr vs phi that may be an indication of 
image

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: I thought about this more, but are there any instances you see where there would be a wrap around problem? Since I take the absolute value of all phi's first, I feel like this should be fine? If you have one hit near pi and the next hit near -pi for example, I take the absolute value of both first so you would get the correct small separation.

Comment (slava77) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: The problem is the continuity is easily violated around 0 and π.
it could be `abs` is a problem around 0, e.g. `0,0.1,-0.15,0.2` is not a natural pattern compared to its abs value.
Or around π e.g. `π-0.1,π,π-0.1` is not good but `π-0.1,π,-π+0.1` is OK.

**Comment by slava77** [comment]
image
CMSSW vs displacement looks good

**Comment by GNiendorf** [comment]
> CMSSW vs displacement looks good

Wow, looks great. Nice decrease in the FR+DR.

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h (hunk 1 above)
Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: I'm finding that using the signed differences results in a bit worse performance. Given that this PR is good-to-go, and this issue also involves the other variables (signed difference of eta for ex.) can we make an issue for this and address it in another PR?

Comment (slava77) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: just to be sure, was the test to use e.g. `deltaPhi(anchorPhi[mdIndex2], anchorPhi[mdIndex1])` instead of what it is now `abs(anchorPhi[mdIndex2])-abs(anchorPhi[mdIndex1])`, or something else?

Numerically this should affect only the small regions near 0 and π. Is the performance getting worse there or in some other regions?

Comment (slava77) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: > deltaPhi(anchorPhi[mdIndex2], anchorPhi[mdIndex1])

in terms of our methods names `phi_mpi_pi(anchorPhi[mdIndex2]-  anchorPhi[mdIndex1])`

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: I implemented my own delta phi just to be sure that I used the same function during training (see below), but I don't think the distributions are going to be exactly the same, right? If both phi values are positive then yes of course, but if both are negative and you have for example -20 degrees to -30 degrees then you would get -10 degrees for deltaPhi(phi2, phi1) vs. positive 10 degrees for abs(-30) - abs(-20).

Screenshot 2024-11-07 at 7 55 02 PM

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:65]: @slava77 I figured the issue out I think. Right now, I'm normalizing the phi input variables by dividing by pi, but when I moved to the signed deltas I stopped normalizing them because their distributions were already small. Removing those normalizations caused the DNN to rely on the phi variables more which reduced the displaced track efficiency slightly at the improvement of a lower FR. If I boosted the displaced track weight up more in-turn maybe it would have sorted this out, but I'm just going to leave the normalizations there. I'll add the update commit tomorrow.

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h]: Fixed, had to increase the weight to 6x either way.

**Comment by GNiendorf** [comment]
Curious if this works still.

/run all lowpt

**Comment by GNiendorf** [comment]
> There was a problem while building and running with CMSSW. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/11756823150).

At least the standalone works still, looks nice.

**Review by slava77 — APPROVED** [review]
